docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,185 @@ With this in mind, we will tackle the migration path via two main areas, txs and
-`Message` type. The `Transaction` message type must implement the SDK's `Tx` interface.
+#### `SIGN_MODE_DEFAULT`
 
-Example:
+The default signing behavior is to sign the raw `TxBody` bytes as broadcast over
+the wire. This has the advantages of:
 
-```protobuf
-// app/codec/codec.proto
+* requiring the minimum additional client capabilities beyond a standard protocol
+buffers implementation
+* leaving effectively zero holes for transaction malleability (i.e. there are no
+subtle differences between the signing and encoding formats which could 
+potentially be exploited by an attacker)
 
-message Transaction {
-  cosmos_sdk.x.auth.v1.StdTxBase base = 1;
-  repeated Message               msgs = 2;
+In order to sign in the default mode, clients take the following steps:
+
+1. Encode `TxBody`
+2. Sign `SignDocRaw`
+
+The raw encoded `TxBody` bytes are encoded into `SignDocRaw` below so that the
+encoded body bytes exactly match the signed body bytes with no need for
+["canonicalization"](https://github.com/regen-network/canonical-proto3) of that
+message.
+
+```proto
+// types/types.proto
+message SignDocRaw {
+    bytes body_bytes = 1;
+    string chain_id = 2;
+    uint64 account_number = 3;
+    uint64 account_sequence = 4;
 }
 ```
 
-Note, the `Transaction` type includes `StdTxBase` which will be defined by the SDK
-and includes all the core field members that are common across all transaction types.
-Developers do not have to include `StdTxBase` if they wish, so it is meant to be
-used as an auxiliary type.
+The only requirements are that the client _must_ encode `SignDocRaw` itself canonically.
+ This means that:
+ 
+* all of the fields must be encoded in order
+* default values (i.e. empty/zero values) must be omitted 
 
-### Signing
+If a protobuf implementation does not by default encode `SignDocRaw` canonically,
+the client _must_ manually encode `SignDocRaw` following the guidelines above.
 
-Signing of a `Transaction` must be canonical across clients and binaries. In order
-to provide canonical representation of a `Transaction` to sign over, clients must
-obey the following rules:
+Again, it does not matter if `TxBody` was encoded canonically or not.
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
-  - Default must be stripped from the output!
-  - JSON keys adhere to their Proto-defined field names.
-- Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
-  - This spec should be trivial to interpret and implement in any language.
+Note that in order to decode `SignDocRaw` for displaying contents to users,
+the regular `SignDoc` type should be used.
 
-```Protobuf
-// app/codec/codec.proto
+3. Broadcast `TxRaw`
 
-message SignDoc {
-  StdSignDocBase base = 1;
-  repeated Message msgs = 2;
+In order to make sure that the signed body bytes exactly match the encoded body
```

Comment:
Why not just broadcast `TxBody`?

Reply:
I'm not sure I understand...

Reply:
Instead of broadcasting `TxRaw`, broadcast `TxBody`.

Reply:
`TxBody` doesn't have the signatures. Do you mean `Tx`? `bytes(TxRaw)` should `==` `bytes(Tx)`

Reply:
Sorry, yes, the broadcast should accept a `Tx`.